### ISSUE TITLE
[Feature request] define axis in 'tf.unique()' and 'tf.unique_with_counts'

### ISSUE THREAD

**Issue by Hoeze** [tensorflow/tensorflow#15644]
Hi, just a small feature request:
It would be cool if one could directly
* use 'tf.unique()' and 'tf.unique_with_counts' in n-dimensional arrays
* define an axis along which 'tf.unique()' and 'tf.unique_with_counts' are applied

**Comment by yongtang** [comment]
The `axis` support for `tf.unique` was added in PR #12952 though it has not been enabled in python yet, as it needs to go through the API deprecation process of 3 weeks. I think it could be enabled by now. Will submit a PR shortly.

**Comment by asimshankar** [comment]
Marking as contributions welcome since @yongtang has a PR to address this. Thanks!

**Comment by aseem-hegshetye** [comment]
I have created a workaround for tf.unique(2D_tensor). Pass any 2D tensor to the below function and it will return a unique 2D tensor.

```
def tf_unique_2d(x):
    x_shape=x.get_shape() #(3,2)
    x1=tf.tile(x,[1,x_shape[0]]) #[[1,2],[1,2],[1,2],[3,4],[3,4],[3,4]..]
    x2=tf.tile(x,[x_shape[0],1]) #[[1,2],[1,2],[1,2],[3,4],[3,4],[3,4]..]

    x1_2=tf.reshape(x1,[x_shape[0]*x_shape[0],x_shape[1]])
    x2_2=tf.reshape(x2,[x_shape[0]*x_shape[0],x_shape[1]])
    cond=tf.reduce_all(tf.equal(x1_2,x2_2),axis=1)
    cond=tf.reshape(cond,[x_shape[0],x_shape[0]]) #reshaping cond to match x1_2 & x2_2
    cond_shape=cond.get_shape()
    cond_cast=tf.cast(cond,tf.int32) #convertin condition boolean to int
    cond_zeros=tf.zeros(cond_shape,tf.int32) #replicating condition tensor into all 0's

    #CREATING RANGE TENSOR
    r=tf.range(x_shape[0])
    r=tf.add(tf.tile(r,[x_shape[0]]),1)
    r=tf.reshape(r,[x_shape[0],x_shape[0]])

    #converting TRUE=1 FALSE=MAX(index)+1 (which is invalid by default) so when we take min it wont get selected & in end we will only take values <max(indx).
    f1 = tf.multiply(tf.ones(cond_shape,tf.int32),x_shape[0]+1)
    f2 =tf.ones(cond_shape,tf.int32)
    cond_cast2 = tf.where(tf.equal(cond_cast,cond_zeros),f1,f2) #if false make it max_index+1 else keep it 1

    #multiply range with new int boolean mask
    r_cond_mul=tf.multiply(r,cond_cast2)
    r_cond_mul2=tf.reduce_min(r_cond_mul,axis=1)
    r_cond_mul3,unique_idx=tf.unique(r_cond_mul2)
    r_cond_mul4=tf.subtract(r_cond_mul3,1)

    #get actual values from unique indexes
    op=tf.gather(x,r_cond_mul4)

    sess=tf.Session()
    return (op)
```

```
import tensorflow as tf
ip=tf.constant([[1,2,1],[3,4,1],[5,6,1],[1,2,1]])
op=tf_unique_2d(ip)
print(sess.run(op))
#op = [[1,2,1],[3,4,1],[5,6,1]]
```

**Comment by jvishnuvardhan** [comment]
Automatically closing this out since I understand it to be resolved by the PR #12952 (merged already), but please let me know if I'm mistaken.Thanks!

**Comment by sp-ananth** [comment]
> * use 'tf.unique()' and 'tf.unique_with_counts' in n-dimensional arrays

Hello, is there still no support for tf unique with multi-dimensional arrays?

**Comment by TianrenWang** [comment]
Ya I also need to use unique with axis. Please fix this as soon as possible.

**Comment by nbro** [comment]
@jvishnuvardhan You should clearly re-open this issue. This has not yet been solved in TF 2.1. The Python API does not seem to support this.

**Comment by jvishnuvardhan** [comment]
@nbro Reopening this. Thanks!

**Comment by zaqqwerty** [comment]
Any progress on this?  My use case is to count unique bitstrings in a 2-D tensor.  Example input tensor:
```python
tf.Tensor(
[[ True False  True]
 [ True False  True]
 [ True False  True]
 [False False False]
 [False False False]], shape=(5, 3), dtype=bool)
```
which should return
```python
y ==> [[True, False, True], [False, False, False]]
idx ==> [0, 0, 0, 1, 1]
count ==> [3, 2]
```

**Comment by spate141** [comment]
Lordy! Can we please have the `tf.unique` with axis!? It's been **964 days** since this issue was created.

**Comment by shariq-audiofocus** [comment]
+1

**Comment by shariq-audiofocus** [comment]
@spate141 - one workaround (no gradients available) - 
```
x.shape  # (100, 2)
unique_results = tf.raw_ops.UniqueV2(x=not_unique, axis=0).y
unique_results.shape  # (89, 2)
```

See here: 
https://www.tensorflow.org/api_docs/python/tf/raw_ops
https://www.tensorflow.org/api_docs/python/tf/raw_ops/UniqueV2

**Comment by rmothukuru** [comment]
@Hoeze,
Sorry for the delayed response. It looks like [UniqueV2 ](https://www.tensorflow.org/api_docs/python/tf/raw_ops/UniqueV2) and [UniqueWithCountsV2](https://www.tensorflow.org/api_docs/python/tf/raw_ops/) has **`axis`** argument. Can you please confirm if this is what you are looking for? Thanks!

**Comment by Hoeze** [comment]
Thanks @rmothukuru, this looks indeed like what I was looking for.
I remember that the UniqueV2 and similar functions were not accessible via the normal Tensorflow API, but unfortunately I did not try tensorflow for years now.

Since there are many other people interested in this feature, maybe someone of you could have a try?

**Comment by rmothukuru** [comment]
@Hoeze,
Thank you for your response. Sure, I will wait for the community to respond about it.

**Comment by pesekon2** [comment]
Commenting just to reopen this issue because I believe that having `tf.unique()` working only for 1D tensor is an avoiding limitation.
